README.md
repo-backend
============

Para obtener los datos de reprepro todo se ejecuta desde $REPREPRO_BASE_DIR

## Listado de distros 
<pre>
$REPREPRO_BASE_DIR/conf/distributions
</pre>
Se puede parsear el archivo con algo tipo ConfigParser de python, para tenes toda la info de cada distro
o solo tener los nombre con:

cmd:
<pre>
grep Codename $REPREPRO_BASE_DIR/conf/distributions
</pre>
output:
<pre>
Codename: brisa
Codename: brisa-updates
Codename: brisa-proposed
Codename: pampero-proposed
Codename: pampero
Codename: pampero-updates
Codename: torbellino
Codename: 1.1
Codename: 2.0
Codename: 2.1
Codename: 2.2
Codename: 3.0
Codename: mate-brisa
Codename: mate-pampero
Codename: mate-torbellino
Codename: sud
Codename: sud-updates
Codename: sud-proposed
Codename: zonda
Codename: zonda-updates
Codename: experimental
Codename: 3.1
Codename: 3.2
</pre>
## Listado de paquetes por distro

reprepro list <distro>

cmd:
<pre>
reprepro list torbellino 
</pre>

output:
<pre>
torbellino|main|i386: touchegg 1.1.1-2
torbellino|main|i386: touchegg-gui 1.1.1-2
torbellino|main|i386: turtleart 170-1huayra1
torbellino|main|i386: unetbootin 608-1
torbellino|main|i386: unetbootin-translations 608-1
torbellino|main|i386: visita 0.5.25
torbellino|main|i386: wari 2.2-7
torbellino|main|i386: xul-ext-downloadhelper 4.9.9-2
torbellino|main|i386: xul-ext-spellchecker-es-ar 2.5-1
torbellino|main|i386: zamba 2.2-7
u|torbellino|main|i386: huayra-archive-keyring-udeb 1.7
torbellino|main|amd64: albatross-gtk-theme 1.9.4-0huayra1
torbellino|main|amd64: anagramarama-data 0.4-9
torbellino|main|amd64: apicaro 1:1.0-2
torbellino|main|amd64: artomico 0.1.0-5
torbellino|main|amd64: ayni 0.3-huayra1
torbellino|main|amd64: babiloo-dicts-spanish 1.0-2
torbellino|main|amd64: base-files 8.5~huayra2
torbellino|main|amd64: bluebird-gtk-theme 1.9.4-0huayra1
</pre>

## Datos de un paquete en todas las distros

reprepro ls <package>

cmd:
<pre>
reprepro ls huayra-desktop
</pre>

output:
<pre>
huayra-desktop | 2.74 |            brisa | source
huayra-desktop | 2.96.2 |    brisa-updates | source
huayra-desktop | 2.96.2 |   brisa-proposed | source
huayra-desktop | 2.161.4 | pampero-proposed | source
huayra-desktop | 2.146 |          pampero | source
huayra-desktop | 2.161.4 |  pampero-updates | source
huayra-desktop | 4.0.8 |       torbellino | source
huayra-desktop | 2.96 |              1.1 | source
huayra-desktop | 2.146 |              2.0 | source
huayra-desktop | 2.161.4 |              2.1 | source
huayra-desktop | 2.161.4 |              2.2 | source
huayra-desktop | 3.2.9 |              3.0 | source
huayra-desktop | 3.2.9 |              sud | source
huayra-desktop | 3.2.37 |      sud-updates | source
huayra-desktop | 3.2.37 |     sud-proposed | source
huayra-desktop | 4.0.8 |            zonda | source
huayra-desktop | 3.2.25 |              3.1 | source
huayra-desktop | 3.2.37 |              3.2 | source
</pre>


debian/changelog
repo-backend (0.2.3~beta2) unstable; urgency=medium

  * Fix for node versions that don't expose mtimeMs in fs.Stats

 -- Ignacio Losiggio <iglosiggio@gmail.com>  Mon, 19 Jun 2017 15:39:05 -0300

repo-backend (0.2.3~beta) unstable; urgency=medium

  * Modificado el sistema para la recarga del repo
  * Agregado logging al ejecutar la tarea repo_load_files
  * 

 -- Ignacio Losiggio <iglosiggio@gmail.com>  Mon, 19 Jun 2017 14:59:27 -0300

repo-backend (0.2.2) unstable; urgency=medium

  * Trabajo inicial en la búsqueda de paquetes

 -- Ignacio Losiggio <iglosiggio@gmail.com>  Wed, 19 Apr 2017 16:11:50 -0300

repo-backend (0.2.1) unstable; urgency=medium

  * Cambios menores en el log, versión para probar que funcione modificar el
    repositorio

 -- Ignacio Losiggio <iglosiggio@gmail.com>  Tue, 06 Dec 2016 11:46:47 -0300

repo-backend (0.2) unstable; urgency=medium

  * Cuando ocurre algún cambio en el repositorio se marca como dirty y se espera
    al job de parsear los archivos. Debería evitar el bug de versiones vacías al
    recargar la data y consumo innecesario de la máquina al duplicarse los eventos

 -- Ignacio Losiggio <iglosiggio@gmail.com>  Mon, 05 Dec 2016 15:01:25 -0300

repo-backend (0.1) unstable; urgency=medium

  * Initial release

 -- Fernando Toledo <ragnarok@docksud.com.ar>  Wed, 12 Oct 2016 14:12:01 -0300


config.test.js
var config = {
	config: function config(name, default_value) {
		this[name] = process.env[name] || default_value;
		return this;
	}
};

config
.config('API_PORT', 8080)
.config('LOG', true)
.config('LOG_EXEC_OUTPUT', false) /* No tiene efecto aún */
.config('REPREPRO_BASE_DIR')
.config('REPO_DISTS_DIR', config.REPREPRO_BASE_DIR + '/dists')
.config('REPO_POOL_DIR', config.REPREPRO_BASE_DIR + '/pool')
.config('LOGFILE')
.config('PIDFILE')
.config('LOAD_INTERVAL', 5000);

config.reprepro = {
	package_versions: 'cat test_data/package_versions',
	distro_list: 'cat test_data/distro_list',
	search_package_files: 'cat test_data/repo_package_files',
	search_source_files: 'cat test_data/repo_source_files',
	list_all_packages: 'cat test_data/repo_package_list'
};

config.prefered_distro = 'torbellino';

module.exports = config;


debug.js
var stacktrace = require('stacktrace-js');

var createWriteStream = require('fs').createWriteStream;
var format = require('util').format;
var basename = require('path').basename;
var config = require('./config.js');
var enabled = config.LOG;
var logfile = config.LOGFILE;

var debug_console;

function create_logfile_console(logfile) {
	var write_stream = createWriteStream(logfile);
	return new console.Console(write_stream, write_stream);
}

function dummy() {
}

debug_console = logfile ? create_logfile_console(logfile) : console;

if(enabled) {
	module.exports = function gen_logger(namespace) {

		function logger() {
			var now = Date();
			var args = Array.prototype.slice.call(arguments);
			var stack = stacktrace.getSync();
			var log_format = "[%s/%s/%s %s:%s] %s";
			var caller = stack[1]; /* stack[0] es logger */
			var f = caller.functionName;
			var file = basename(caller.fileName); /* Muestro únicamente el nombre del archivo */
			var line = caller.lineNumber;
			var args_format = args[0];
			var args_args = args.slice(1);
			var string = format(log_format, now, namespace, f, file, line, args_format);

			if(args_args.length > 0) {
				debug_console.log.apply(debug_console, [string].concat(args_args));
			} else {
				debug_console.log(string);
			}
		}

		return logger;
	};
} else {
	module.exports = function dummy_logger() {
		return dummy;
	};
}

index.js
/* Imports */
var Promise = require('es6-promise');
var express = require('express');
var exec = require('child_process').exec;
var utils = require('./utils');
var util = require('util');
var debian_packages = require('./debian_packages');
var config = require('./config');
var debug = require('./debug')('webapp');
var node_error = require('./debug')('NODE-ERROR');
var repo = {};

var format = util.format;
var format_map = utils.format_map;
var object_map = utils.object_map;
var object_values = utils.object_values;

/* Variables */
var app = express();

function sanitize_input(req) {
	function strip_illegal_chars(str) {
		return str.replace(/[^a-z0-9-+.]/g, '');
	}
	return object_map(req, strip_illegal_chars);
}

process.on('uncaughtException', function evitar_que_explote_todo(error) {
	node_error('Se produjo una excepción que no fué capturada (casi cierra el proceso)');
	console.log(error.stack || error.message);
});

/* Agrego los headers de CORS */
app.use(function add_cors_headers(req, res, next) {
	res.header("Access-Control-Allow-Origin", "*");
	res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
	next();
});

app.get('/packages/search/:query', function get_package_search(req, res) {
	var params = sanitize_input(req.params);
	var search = new RegExp('.*' + params.query + '.*', 'g');
	var results = repo.package_list.list.match(search) || [];

	var query_results = results.map(function augument_result(title) {
		var package = repo.binaries.get_any(title, config.prefered_distro) || repo.sources.get_any(title, config.prefered_distro) || {};
		var description = package.Description || {};

		return {
			title: title,
			description: description['Short-Description'],
		};
	});

	res.send({ results: query_results });
});

/*
 * Salida esperada:
 *
 * {
 *   package: 'huayra-tpm',
 *   versions: [
 *     {
 *       version: '2.0.0.116-7',
 *       distribution: 'brisa',
 *       archs: ['i386', 'source']
 *     },
 *     {
 *       version: '2.0.0.117-3',
 *       distribution: 'torbellino',
 *       archs: ['i386', 'amd64', 'source']
 *     },
 *   ]
 * }
 */
app.get('/packages/:package', function get_package_versions(req, res) {
	var params = sanitize_input(req.params);
	var cmdline = format_map(config.reprepro.package_versions, req.params);

	debug(req.method, req.url);

	exec(cmdline, function exec_package_versions(execerror, stdout, stderr) {
		var salida = stdout
			.toString()
			.replace(/ /g, '')   /* Quito todos los espacios  */
			.replace(/^u\|/m, '') /* la marca de paquetes udeb */
			.replace(/\n$/g, '');  /* y el salto de línea final */
		var versiones;

		function add_version(versions, text) {
			var data = text.split('|');

			return {
				package: data[0],
				versions: versions.versions.concat({
					version: data[1],
					distribution: data[2],
					archs: data[3].split(',')
				})
			};
		}

		/* Si la salida es vacía no se encontró el paquete */
		if(salida === '') {
			var error = {
				code: 404,
				message: 'No se encontró el paquete \'' + params.package + '\'',
				params: params,
				stderr: stderr.toString()
			};

			res.status(404);
			res.send(error);
			debug('NOT-FOUND:', req.method, req.url, error.message);

			return;
		}

		versiones = salida
			.split('\n')
			.reduce(add_version, { versions: [] });

		res.send(versiones);
	});
});

app.get('/packages/:distro/:package', function get_package_info(req, res) {
	var params = sanitize_input(req.params);
	var distro = params.distro;
	var package_name = params.package;
	var package;
	var error;

	debug(req.method, req.url);

	package = repo.binaries.get(distro, package_name);

	if(package === undefined) {
		error = {
			code: 404,
			message: format('No se encontró el paquete binario \'%s\' en la distro \'%s\'', package_name, distro),
			params: params
		};

		res.status(404);
		res.send(error);
		debug('NOT-FOUND:', req.method, req.url, error.message);

		return;
	}

	res.send(package);
});

app.get('/sources/:distro/:package', function get_source_info(req, res) {
	var params = sanitize_input(req.params);
	var distro = params.distro;
	var package_name = params.package;
	var package;
	var error;

	debug(req.method, req.url);

	package = repo.sources.get(distro, package_name);

	if(package === undefined) {
		error = {
			code: 404,
			message: 'No se encontró el paquete source \'' + package_name + '\' en la distro \'' + distro + '\'',
			params: params
		};

		res.status(404);
		res.send(error);
		debug('NOT-FOUND:', req.method, req.url, error.message);

		return;
	}

	res.send(package);
});

app.get('/distributions', function get_distro_list(req, res) {
	var cmdline = config.reprepro.distro_list;
	var distributions;

	debug(req.method, req.url);

	exec(cmdline, function exec_distro_list(error, stdout, stderr) {

		var salida = stdout.toString();

		function add_distro(text) {

			function add_field(distro, text) {
				var field = text.split(': ');
				distro[field[0]] = field[1];
				return distro;
			}

			return text
				.split('\n')
				.reduce(add_field, {});
		}

		distributions = salida
			.split('\n\n')
			.map(add_distro);

		res.send(distributions);
	});
});

function extract_data(package) {

	function extract_versions(version) {
		return {
			Version: version.Version,
			Architecture: version.Architecture
		};
	}

	return {
		Package: package.Package,
		Description: package.Description['Short-Description'],
		Component: package.Component,
		versions: package.versions.map(extract_versions)
	};
}


app.get('/distributions/:distro/packages', function get_distro_packages(req, res) {
	var params = sanitize_input(req.params);
	var packages = repo.binaries.get_distro(params.distro);
	var package_list = object_values(packages).map(extract_data);

	debug(req.method, req.url);

	/* Si la salida es vacía no se encontró el paquete */
	if(Object.keys(packages).length === 0) {
		var error = {
			code: 404,
			message: 'No existe la distribución \'' + params.distro + '\'',
			params: params
		};

		res.status(404);
		res.send(error);
		debug('NOT-FOUND:', req.method, req.url, error.message);

		return;
	}

	res.send(package_list);
});

app.get('/distributions/:distro/sources', function get_distro_sources(req, res) {
	var params = sanitize_input(req.params);
	var packages = repo.sources.get_distro(params.distro);
	var package_list = object_values(packages).map(extract_data);

	debug(req.method, req.url);

	/* Si la salida es vacía no se encontró el paquete */
	if(Object.keys(packages).length === 0) {
		var error = {
			code: 404,
			message: 'No existe la distribución \'' + params.distro + '\'',
			params: params
		};

		res.status(404);
		res.send(error);
		debug('NOT-FOUND:', req.method, req.url, error.message);

		return;
	}

	res.send(package_list);
});

function load_packages() {
	function binaries_loaded(binaries) {
		repo.binaries = binaries;

		debug('Cargados los binarios');
	}

	function sources_loaded(sources) {
		repo.sources = sources;

		debug('Cargados los sources');
	}

	function packages_listed(packages) {
		repo.package_list = packages;

		debug('Cargada la lista de paquetes');
	}

	Promise.all([
		debian_packages.init_binaries().then(binaries_loaded),
		debian_packages.init_sources().then(sources_loaded),
		debian_packages.load_package_list().then(packages_listed)
	]).then(function start_app() {
		app.listen(config.API_PORT, function start_server() {
		  console.log('repo-backend listening on port', config.API_PORT);
		});
	});

	debug('Cargando binarios y sources');
}

load_packages();

if(config.PIDFILE) {
	debug("Creando pidfile (%s)", config.PIDFILE);

	require('fs').writeFile(config.PIDFILE, process.pid);

	process.on('exit', function remove_pidfile() {
		debug("Borrando pidfile (%s)", config.PIDFILE);
		require('fs').unlinkSync(config.PIDFILE);
	});

	process.on('SIGTERM', function recv_sigterm() {
		debug("Recibí un sigterm, buenas noches!");
		process.exit();
	});

	process.on('SIGINT', function recv_sigint() {
		debug("Recibí un sigint, buenas noches!");
		process.exit();
	});
}


utils.js
module.exports.array_iterator = function array_iterator(array) {
	array = array_clone(array).reverse();

	return function pop_value() {
		return array.pop();
	};
};

module.exports.array_clone = function array_clone(array) {
	return array.concat([]);
};

module.exports.format_map = function format_map(string, map) {
	function get_param(match, param) {
		return map[param];
	}

	return string.replace(/<([^<>]+)>/g, get_param);
};

module.exports.object_map = function object_map(obj, f) {
	var rval = {};

	for(var key in obj) {
		rval[key] = f(obj[key], key, obj);
	}
	return rval;
};

module.exports.object_valuemap = function object_valuemap(obj, f) {
	var rval = [];

	for(var key in obj) {
		rval.push(f(obj[key], key, obj));
	}
	return rval;
};

module.exports.regex_fold = function regex_fold(regex, string, f, initial_value) {
	var match = regex.exec(string);
	var accumulator = initial_value;

	while(match) {
		accumulator = f(accumulator, match);
		match = regex.exec(string);
	}
	return accumulator;
};

module.exports.object_values = function object_values(obj) {
	function get_field(field) {
		return obj[field];
	}

	return Object.keys(obj).map(get_field);
};

module.exports.object_filter = function object_filter(obj, cb) {
	var rval = {};

	for(var key in obj) {
		if(cb(obj[key], key, obj)) {
			rval[key] = obj[key];
		}
	}

	return rval;
};

module.exports.object_merge = function object_merge(dest, orig) {
	for(var key in orig) {
		dest[key] = orig[key];
	}

	return dest;
};

module.exports.object_reduce = function object_reduce(obj, operation, initial_value) {
	var rval = initial_value;

	for(var key in obj) {
		rval = operation(rval, obj[key], key, obj);
	}

	return rval;
};

module.exports.get_field = function get_field(field) {
	return function getter(obj) {
		return obj[field];
	};
};

module.exports.Waiter = function Waiter(cb, cb_this, args) {
	if(!(this instanceof Waiter)) {
		return new Waiter(cb, cb_this, args);
	}

	if(cb) {
		this.set_cb(cb, cb_this, args);
	}

	this._waits = [];
};

module.exports.Waiter.prototype = {
	/* Returns a callback that waits */
	wait: function wait() {
		var self = this;
		var wait = {
			done: false,
			called: function called(called) {
				wait.done = true;
				wait.arguments = arguments;

				if(self.cb && self._waits.every(is_fullfiled)) {
					self.cb(self._waits);
				}
			}
		};

		function is_fullfiled(wait) {
			return wait.done;
		}

		this._waits.push(wait);

		return wait.called;
	},

	set_cb: function set_cb(cb, cb_this, args) {

		function is_fullfiled(wait) {
			return wait.done;
		}

		if(cb_this || args) {
			this.cb = cb.bind.apply(cb, [cb_this].concat(args));
		} else {
			this.cb = cb;
		}

		if(this._waits.every(is_fullfiled)) {
			this.cb(this._waits);
		}
	}
};

config.prod.js
var config = {
	config: function config(name, default_value) {
		this[name] = process.env[name] || default_value;
		return this;
	}
};

config
.config('API_PORT', 8080)
.config('LOG', true)
.config('LOG_EXEC_OUTPUT', false) /* No tiene efecto aún */
.config('REPREPRO_BASE_DIR')
.config('REPO_DISTS_DIR', config.REPREPRO_BASE_DIR + '/dists')
.config('REPO_POOL_DIR', config.REPREPRO_BASE_DIR + '/pool')
.config('LOGFILE')
.config('PIDFILE')
.config('LOAD_INTERVAL', 600000); /* 10 minutos */

config.reprepro = {
	package_versions: 'reprepro ls "<package>"',
	distro_list: 'cat ' + config.REPREPRO_BASE_DIR + '/conf/distributions',
	search_package_files: 'find "' + config.REPO_DISTS_DIR + '" -iname Packages.gz',
	search_source_files: 'find "' + config.REPO_DISTS_DIR + '" -iname Sources.gz',
	list_all_packages: 'find ' + config.REPO_POOL_DIR + ' -type f | awk -F/ \'{ gsub(/_.*$/, "", $NF); print $NF}\' | sort | uniq'
};

module.exports = config;


debian_packages.js
/* Imports */
var Promise = require('es6-promise');
var fs = require('fs');
var zlib = require('zlib');
var exec = require('child_process').exec;
var config = require('./config');
var utils = require('./utils');
var debug = require('./debug')('debian_parsing');

var format = utils.format_map;
var regex_fold = utils.regex_fold;
var get_field = utils.get_field;

/* Intervalo en el que node hace polling de los archivos, mínimo 5007 y máximo ~5 segundos antes del LOAD_INTERVAL */
var watchFile_interval = Math.max(config.LOAD_INTERVAL - 5007, 5007);

/* Diccionario para parsear fields específicos */
var FIELD = {};

function init_parser(files) {
	var repo = { dirty: true };

	function watch_file(filename) {
		var filewatch;
		var distro_regex = /\/([^/]+)\/[^/]+\/[^/]+\/[^/]+$/;
		var distro_match = distro_regex.exec(filename);
		var distro = distro_match && distro_match[1] ? distro_match[1] : 'unknown';

		function mark_dirty(stats_now, last_stats) {
			var usable_filename = filename.replace(config.REPO_DISTS_DIR, '');

			/* Date.valueOf() returns the javascript timestamp (miliseconds since unix epoch) */
			if(stats_now.mtime.valueOf() != last_stats.mtime.valueOf()) {
				debug('Se modificó [%s]', usable_filename);

				repo.dirty = true;
			}
		}

		filewatch = {
			filename: filename,
			distro: distro,
			watch: fs.watchFile(filename, { interval: watchFile_interval }, mark_dirty)
		};

		return filewatch;
	}

	repo.watches = files.map(watch_file);
	repo.get = repo_get.bind(null, repo);
	repo.get_distro = repo_get_distro.bind(null, repo);
	repo.get_any = repo_get_any.bind(null, repo);
	repo.contents = {};
	repo.loadInterval = setInterval(repo_load_files, config.LOAD_INTERVAL, repo);

	/* Cargo los archivos */
	repo_load_files(repo);

	return repo;
}

function repo_load_files(repo) {
	var read_files_promises;

	debug(repo.dirty
		? 'Hubo cambios en el repositorio, recargando...'
		: 'No se registran cambios en el repositorio');

	if(!repo.dirty) {
		return;
	}

	/* Recibe un watch y devuelve una Promesa que resuelve a {watch: watch, packages: [paquetes]} */
	function read_to_repo(watch) {
		function add_metadata(packages) {
			return { watch: watch, packages: packages };
		}

		return read_file_promise(watch.filename).then(add_metadata);
	}

	/* Genero un mapa { <distro> => [paquetes] } */
	function divide_distros(repo, data) {
		var distro = data.watch.distro;

		if(repo[distro]) {
			repo[distro] = repo[distro].concat(data.packages);
		} else {
			repo[distro] = data.packages;
		}

		return repo;
	}

	/* Genero un mapa { <distro> => { <paquete> => {contenidos} } } */
	function create_distro_dictionary(packages) {
		packages = packages || []; /* Si por alguna razón no hay packages hago un array vació en lugar de undefined */
		return packages.reduce(fold_packages, {});
	}

	function unify_data(data) {
		var distros;

		debug('Unificando los datos de cada archivo');

		distros = data.reduce(divide_distros, {});
		repo.contents = utils.object_map(distros, create_distro_dictionary);
		repo.dirty = false;
	}

	read_files_promises = repo.watches.map(read_to_repo);
	Promise.all(read_files_promises).then(unify_data);
}

function read_file_promise(filename) {
	return new Promise(read_file.bind(null, filename));
}

function read_file(filename, cb) {
	var usable_filename = filename.replace(config.REPO_DISTS_DIR, '');

	function parse_file(error, data) {
		var component_match = /\/([^/]+)\/[^/]+\/[^/]+$/.exec(filename); /* Matcheo el directorio correspondiente al componente */
		var component = component_match && component_match[1] ? component_match[1] : 'unknown'; /* Si no matchié algo con sentido no se cuál es el componente */
		var packages_text = data.toString();
		var package_list = parse_packages(packages_text, component); /* Parseo este archivo */

		cb(package_list, filename);
	}

	function gunzip(error, data) {
		zlib.gunzip(data, parse_file);
	}

	debug('Leyendo el archivo %s', usable_filename);

	fs.readFile(filename, gunzip);
}

function repo_get(repo, distro, package) {
	var distro_repo;

	distro_repo = repo.contents[distro] || {};

	return distro_repo[package];
}

function repo_get_distro(repo, distro) {
	return repo.contents[distro] || {};
}

function repo_get_any(repo, package_name, preferences) {
	var package;

	if(preferences instanceof Array) { /* Si es un array supongo que es un array de distros */
		package = preferences.find(function find_package(distro) {
			return repo.get(distro, package_name);
		});
	} else if(preferences instanceof String) { /* Si es un string supongo que es la distro */
		package = repo.get(preferences, package_name);
	}

	if(package !== undefined) {
		return package;
	}

	for(var distro in repo.contents) {
		package = repo.get(distro, package_name);

		if(package !== undefined) {
			return package;
		}
	}
}

function init_repo(search_repo_files_cmdline, cb) {
	var cmdline = search_repo_files_cmdline;

	function exec_search_files(error, stdout, stderr) {
		var files = stdout
			.toString()
			.replace(/\n+$/, '') /* Vuelo el salto de línea final */
			.split('\n');
		var repo = init_parser(files);

		cb(repo);
	}

	exec(cmdline, exec_search_files);
}

function init_binaries() {
	var cmdline = config.reprepro.search_package_files;
	return new Promise(function loaded(accept) {
		init_repo(cmdline, accept);
	});
}

function init_sources() {
	var cmdline = config.reprepro.search_source_files;

	return new Promise(function loaded(accept) {
		init_repo(cmdline, accept);
	});
}

function load_package_list() {
	var cmdline = config.reprepro.list_all_packages;
	var package_list = { dirty: true };

	function load() {
		return new Promise(function loaded(accept) {
			if(package_list.dirty) {
				debug('Lista de paquetes sin actualizar, actualizando');
				exec(cmdline, function all_packages(execerror, packages) {
					debug('Cargados %s nombres en la lista de paquetes', packages.split('\n').length);
					package_list.list = packages;
					package_list.dirty = false;
					accept(package_list);
				});
			}
		});
	}

	fs.watch(config.REPO_POOL_DIR, { recursive: true }, function mark_dirty() {
		package_list.dirty = true;
	});

	setInterval(load, config.LOAD_INTERVAL);

	return load();
}

function parse_packages(text, component) {
	var packages;

	function add_package(text) {
		var field_regex = /.+(\n[ ].*)*/g;
		var package;

		function add_field(package, field_match) {
			var field = field_match[0].split(':');
			var fieldname = field[0];

			package[fieldname] = field
				.slice(1) /* Quito el fieldname */
				.join(':') /* junto todo */
				.replace(/^ /mg, '') /* quito prefijos inútiles */
				.replace(/^\n/g, '');


			/* Si hay una función para parsear este field en particular la uso */
			if(FIELD[fieldname] !== undefined) {
				package[fieldname] = FIELD[fieldname](package[fieldname], fieldname);
			}

			return package;
		}

		package = regex_fold(field_regex, text, add_field, {});
		package.Component = component;

		return package;
	}

	packages = text.split('\n\n').slice(0, -1).map(add_package);

	return packages;
}

function fold_packages(packages, package) {
	/*
	 * Algo feo que hago acá es usar keys en un array de JS,
	 * pero al encodearlo a JSON desaparecen, entonces
	 *                   ¯\_(ツ)_/¯
	 */
	var name = package.Package;
	var arch = package.Architecture;
	var version = {
		Architecture: package.Architecture,
		Version: package.Version,
		Size: package.Size,
		Filename: package.Filename,
		'Installed-Size': package['Installed-Size'],
	};

	if(packages[name]) {
		/* Un paquete all va a aparecer en varios lugares, (o podría existir en varios componentes) */
		if(packages[name].versions[arch] === undefined) {
			packages[name].versions.push(version);
			packages[name].versions[arch] = version;
		}
	} else {
		packages[name] = package;
		packages[name].versions = [version];
		packages[name].versions[arch] = version;

		Object.keys(version).forEach(function unset_key(key) {
			delete packages[name][key];
		});
	}
	return packages;
}

function parse_depends(text) {
	var depends = text
		.replace(/ |\n/g, '') /* Todos los espacios y saltos de línea son inútiles */
		.split(',') /* Separo las dependencias */
		.map(function split_alternatives(depend) { /* Separo las alternativas */
			return depend.split('|');
		})
		.map(function parse_depend(depend_text) { /* Parseo la dependencia y sus alternativas */
			var pattern = /([^(]+)(\((>>|>=|=|<=|<<|<|>)([^)]+)\))?/;
			var matches = depend_text.map(function match_depend(alternative) {
				return pattern.exec(alternative);
			});
			var alternatives = matches.slice(1);
			var depend;

			function build_depend(match) {
				var depend = { Package: match[1] };

				if(match[2]) {
					depend.Version = {
						Text: match[2],
						Relation: match[3],
						Version: match[4]
					};
				}

				return depend;
			}

			depend = build_depend(matches[0]);

			if(alternatives.length > 0) {
				depend.Alternatives = alternatives.map(build_depend);
			}

			return depend;
		});
	depends.Text = text;

	return depends;
}

function parse_description(text) {
	return {
		Text: text,
		'Short-Description': text.split('\n')[0], /* Lo separo por saltos de línea y agarro la primer línea */
		'Long-Description': text.replace(/^[^\n]*\n/, '') /* Vuelo todo lo anterior al primer salto de línea */
	};
}

function split_commas(text, fieldname) {
	var rval = { Text: text };
	rval[fieldname] = text.split(', ');

	return rval;
}

function split_newlines(text, fieldname) {
	var rval = { Text: text };
	rval[fieldname] = text.split('\n');

	return rval;
}

FIELD.Depends = parse_depends;
FIELD.Suggests = parse_depends;
FIELD.Conflicts = parse_depends;
FIELD.Recomends = parse_depends;
FIELD['Pre-Depends'] = parse_depends;
FIELD['Build-Depends'] = parse_depends;
FIELD['Build-Depends-Indep'] = parse_depends;
FIELD.Description = parse_description;
FIELD.Binary = split_commas;
FIELD['Package-List'] = split_newlines;
FIELD['Checksums-Sha1'] = split_newlines;
FIELD['Checksums-Sha256'] = split_newlines;
FIELD.Files = split_newlines;

module.exports = {
	init_parser: init_parser,
	read_file: read_file,
	repo_get: repo_get,
	repo_get_distro: repo_get_distro,
	repo_get_any: repo_get_any,
	init_binaries: init_binaries,
	init_sources: init_sources,
	load_package_list: load_package_list,
	parse_packages: parse_packages,
	parse_depends: parse_depends,
	parse_description: parse_description,
	split_commas: split_commas,
	split_newlines: split_newlines
};


search_ideas/time_infix_search.js
const text = require('fs').readFileSync('paquetes', 'utf8');
const list = text.split('\n');

// Timing: ~50ms
function search(term) {
	var result = [];
	var match;
	const regexp = RegExp(`.*(${term}).*`, 'g');
	while(match = regexp.exec(text)) {
		result.push(match[0]);
	}
	return result;
}

// Timing: ~20ms
function naivesearch(term) {
	return list.filter(function(paquete) {
		return paquete.search(term) != -1
	});
}

// Timing: ~15ms
function matchsearch(term) {
	const regexp = RegExp(`.*${term}.*`, 'g');
	return text.match(regexp);
}
const start = process.hrtime();
search('ssl');
search('huayra-');
console.log(process.hrtime(start)); // Log passed time
const naivestart = process.hrtime();
naivesearch('ssl');
naivesearch('huayra-');
console.log(process.hrtime(naivestart)); // Log passed time
const matchstart = process.hrtime();
matchsearch('ssl');
matchsearch('huayra-');
console.log(process.hrtime(matchstart)); // Log passed time


config.js
var config = {
	config: function config(name, default_value) {
		this[name] = process.env[name] || default_value;
		return this;
	}
};

config
.config('API_PORT', 8080)
.config('LOG', true)
.config('LOG_EXEC_OUTPUT', false) /* No tiene efecto aún */
.config('REPREPRO_BASE_DIR')
.config('REPO_DISTS_DIR', config.REPREPRO_BASE_DIR + '/dists')
.config('REPO_POOL_DIR', config.REPREPRO_BASE_DIR + '/pool')
.config('LOGFILE')
.config('PIDFILE')
.config('LOAD_INTERVAL', 600000); /* 10 minutos */

config.reprepro = {
	package_versions: 'reprepro ls "<package>"',
	distro_list: 'cat ' + config.REPREPRO_BASE_DIR + '/conf/distributions',
	search_package_files: 'find "' + config.REPO_DISTS_DIR + '" -iname Packages.gz',
	search_source_files: 'find "' + config.REPO_DISTS_DIR + '" -iname Sources.gz',
	list_all_packages: 'find ' + config.REPO_POOL_DIR + ' -type f | awk -F/ \'{ gsub(/_.*$/, "", $NF); print $NF}\' | sort | uniq'
};

config.prefered_distro = 'torbellino';

module.exports = config;
